Support new window function `ntile`

lgbo-ustc opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- New Feature



### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...
Support window function `ntile`.
```
insert into test_data values(1,2), (1,3), (1,4), (2,5),(2,6);
select a, b, ntile(2) over (partition by a order by b rows between unbounded preceding and unbounded following ) from test_data;

┌─a─┬─b─┬─ntile(2) OVER (PARTITION BY a ORDER BY b ASC)─┐
│ 1 │ 2 │                                             1 │
│ 1 │ 3 │                                             1 │
│ 1 │ 4 │                                             2 │
│ 2 │ 5 │                                             1 │
│ 2 │ 6 │                                             2 │
└───┴───┴───────────────────────────────────────────────┘
```
### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

lgbo-ustc committed:
add new window function, ntile

lgbo-ustc commented:
There is a related issue, https://github.com/ClickHouse/ClickHouse/issues/28989

lgbo-ustc committed:
update doc and add tests

lgbo-ustc committed:
fixed test case

vdimir commented on tests/queries/0_stateless/01591_window_functions.sql:
```diff
@@ -545,3 +545,7 @@ select count() over (w) from numbers(1) window w as ();
 -- nonexistent parent window
 select count() over (w2 rows unbounded preceding); -- { serverError 36 }
+
+-- ntile
```
I'd suggest create separate test file for this function and cover more cases (different arg types, corner cases with illegal arguments and unsupported frames, etc)

vdimir replied:
Result with Float argument unexpected:
```sql
select
	number % 2 as a,
	number as b,
	ntile(2 :: Float64) over (partition by a order by b)  as ntile
from numbers(10);


┌─a─┬─b─┬─ntile─┐
│ 0 │ 0 │     1 │
│ 0 │ 2 │     2 │
│ 0 │ 4 │     3 │
│ 0 │ 6 │     4 │
│ 0 │ 8 │     5 │
│ 1 │ 1 │     1 │
│ 1 │ 3 │     2 │
│ 1 │ 5 │     3 │
│ 1 │ 7 │     4 │
│ 1 │ 9 │     5 │
└───┴───┴───────┘

```

Same for string: `ntile('2')`

vdimir replied:
Non constant argument probably should be forbidden also
```sql
select number%2, number, ntile(number + 2) over (partition by number % 2 order by number)  as ntile from numbers(10);
```

vdimir commented on src/Processors/Transforms/WindowTransform.cpp:
```diff
@@ -1970,6 +1969,121 @@ struct WindowFunctionRowNumber final : public WindowFunction
+            {
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "ntile's argument must > 0");
+            }
+            buckets = static_cast<UInt64>(n);
```
If we cant to `UInt64`, let's forbid bigger argument types (`(U)Int128/256`), `Decimals`, and `Float`. Or better just allow only `(U)Int8/16/32/64`

vdimir commented on src/Processors/Transforms/WindowTransform.cpp:
```diff
@@ -1970,6 +1969,121 @@ struct WindowFunctionRowNumber final : public WindowFunction
+                current_bucket_capacity += 1;
+                capacity_diff--;
+            }
+            Int64 left_rows = static_cast<Int64>(current_bucket_capacity);
```
Why do we cast to signed?

lgbo-ustc committed:
fixed

lgbo-ustc committed:
fixed

change window specification as (partition by a order by b rows between unbounded preceding and unbounded following)

lgbo-ustc committed:
fixed binary_tidy build failure

vdimir reviewed (approved):
Thank you for making an effort to support this window function!

alexey-milovidov commented:
#46763
